source/darwin/dispatch_queue_event_loop.c

```diff
@@ -0,0 +1,538 @@
+                    &dispatch_loop->synced_task_data.context->scheduling_state.scheduled_services, next_task_time)) {
+                s_try_schedule_new_iteration(contxt, next_task_time);
+            }
+        }
+    }
+
+    aws_mutex_unlock(&contxt->lock);
+    s_scheduled_service_entry_destroy(entry);
+}
+
+// Iteration function that scheduled and executed by the Dispatch Queue API
+static void s_run_iteration(void *context) {
+    struct scheduled_service_entry *entry = context;
+
+    struct dispatch_loop_context *dispatch_queue_context = entry->dispatch_queue_context;
+    aws_mutex_lock(&dispatch_queue_context->lock);
+    struct dispatch_loop *dispatch_loop = entry->dispatch_queue_context->io_dispatch_loop;
+    aws_mutex_unlock(&dispatch_queue_context->lock);
+    if (!dispatch_loop) {
+        s_scheduled_service_entry_destroy(entry);
+        return;
+    }
+
+    if (!begin_iteration(entry)) {
+        s_scheduled_service_entry_destroy(entry);
+        return;
+    }
+
+    aws_event_loop_register_tick_start(dispatch_loop->base_loop);
+    // run the full iteration here: local cross-thread tasks
+
+    while (!aws_linked_list_empty(&dispatch_loop->local_cross_thread_tasks)) {
+        struct aws_linked_list_node *node = aws_linked_list_pop_front(&dispatch_loop->local_cross_thread_tasks);
+        struct aws_task *task = AWS_CONTAINER_OF(node, struct aws_task, node);
+
+        /* Timestamp 0 is used to denote "now" tasks */
+        if (task->timestamp == 0) {
+            aws_task_scheduler_schedule_now(&dispatch_loop->scheduler, task);
+        } else {
+            aws_task_scheduler_schedule_future(&dispatch_loop->scheduler, task, task->timestamp);
+        }
+    }
+
+    aws_mutex_lock(&dispatch_loop->synced_thread_data.thread_data_lock);
+    dispatch_loop->synced_thread_data.current_thread_id = aws_thread_current_thread_id();
+    dispatch_loop->synced_thread_data.is_executing = true;
+    aws_mutex_unlock(&dispatch_loop->synced_thread_data.thread_data_lock);
+
+    // run all scheduled tasks
+    uint64_t now_ns = 0;
+    aws_event_loop_current_clock_time(dispatch_loop->base_loop, &now_ns);
+    aws_task_scheduler_run_all(&dispatch_loop->scheduler, now_ns);
+    aws_event_loop_register_tick_end(dispatch_loop->base_loop);
+
+    aws_mutex_lock(&dispatch_loop->synced_thread_data.thread_data_lock);
+    dispatch_loop->synced_thread_data.is_executing = false;
+    aws_mutex_unlock(&dispatch_loop->synced_thread_data.thread_data_lock);
+
+    end_iteration(entry);
+}
+
+/**
+ * Checks if a new iteration task needs to be scheduled, given a target timestamp. If so, submits an iteration task to
+ * dispatch queue and registers the pending execution in the event loop's list of scheduled iterations.
+ *
+ * If timestamp==0, the function will always schedule a new iteration as long as the event loop is not suspended.
+ *
+ * The function should be wrapped with dispatch_loop->synced_task_data->lock
```

Comment:
This comment does not match the reality.  This function is getting called with the context lock held not the loop lock.